src/network/wait-online/link.c

```diff
@@ -145,6 +152,90 @@ static int link_update_altnames(Link *l, sd_netlink_message *m) {
         return 0;
 }
 
+static int link_update_dns_from_resolv_conf(Link *l) {
+        _cleanup_strv_free_ char **dns = NULL;
+        _cleanup_fclose_ FILE *f = NULL;
+        int r;
+
+        log_link_debug(l, "Attempting to read DNS servers from /etc/resolv.conf");
+
+        f = fopen("/etc/resolv.conf", "re");
+        if (!f)
+                return -errno;
+
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+                const char *a;
+
+                r = read_stripped_line(f, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                if (IN_SET(*line, '#', ';', 0))
+                        continue;
+
+                a = first_word(line, "nameserver");
+                if (!a)
+                        continue;
+
+                for (;;) {
+                        _cleanup_free_ char *word = NULL;
+                        union in_addr_union addr;
+                        int family;
+
+                        r = extract_first_word(&a, &word, NULL, 0);
+                        if (r < 0)
+                                return r;
+                        if (r == 0)
+                                break;
+
+                        r = in_addr_from_string_auto(word, &family, &addr);
+                        if (r < 0)
+                                return r;
+
+                        if (dns_server_address_valid(family, &addr)) {
+                                r = strv_push(&dns, TAKE_PTR(word));
+                                if (r < 0)
+                                        return r;
+                        }
+                }
+        }
+
+        strv_free_and_replace(l->dns, dns);
+
+        return 0;
+}
+
+static int link_update_dns(Link *l) {
+        _cleanup_(sd_varlink_unrefp) sd_varlink *vl = NULL;
+        int r;
+
+        /* If we determine that systemd-resolved is running, then we can keep
+         * things simple and check sd_network_link_get_dns() to determine DNS
+         * servers configured for an interface.
+         *
+         * In the case that systemd-resolved is not in use, let's check
+         * /etc/resolv.conf directly. */
+        r = sd_varlink_connect_address(&vl, "/run/systemd/resolve/io.systemd.Resolve");
+        if (r < 0) {
+                r = link_update_dns_from_resolv_conf(l);
+                if (r < 0)
+                        return r;
+        } else {
+                _cleanup_strv_free_ char **dns = NULL;
+
+                r = sd_network_link_get_dns(l->ifindex, &dns);
+                if (r < 0)
+                        return r;
+
+                strv_free_and_replace(l->dns, dns);
```

Comment:
hmm, i find this a bit sloppy tbh, i.e. running this fallback path once for each interface, and then claiming global DNS was actually local DNS is a bit sloppy, no?

I think we should just propagate an error here if we cannot connect to resolved, and store the inability to talk to resolved as part of the Link object. And then, when determining the online status, read the resolve.conf file once, associated to the Manager object if we see this issue.